File: backend/src/std_cards/models/analytics.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel


class DashboardKpi(BaseModel):
    total_scans: int
    last_30d_scans: int
    unique_cards: int
    active_members: int


class TimelinePoint(BaseModel):
    day: date
    count: int


class RegionStats(BaseModel):
    """Top regions for scans. Prefers city, falls back to country code if city unknown."""

    region: str
    count: int


class DeviceStats(BaseModel):
    device_type: str
    count: int


class TopCard(BaseModel):
    card_id: UUID
    last_name: str
    first_name: str
    membership_no: str
    scans: int


class TopActiveUser(BaseModel):
    card_id: UUID
    last_name: str
    first_name: str
    middle_name: str | None = None
    category_name: str | None = None
    membership_no: str
    scans: int
    top_region: str | None = None
    top_device: str | None = None


class TopActiveUsersResponse(BaseModel):
    items: list[TopActiveUser]
    total: int
    page: int
    page_size: int


class DashboardResponse(BaseModel):
    kpi: DashboardKpi
    by_day: list[TimelinePoint]
    top_regions: list[RegionStats]
    top_devices: list[DeviceStats]
    top_cards: list[TopCard]


class CardAnalytics(BaseModel):
    card_id: UUID
    total_scans: int
    last_scan: datetime | None
    by_day: list[TimelinePoint]
    by_region: list[RegionStats]
    by_device: list[DeviceStats]


class ScanEvent(BaseModel):
    card_id: UUID
    ts: datetime
    ip: str | None = None
    user_agent: str | None = None
    referer: str | None = None

File: frontend/src/lib/api/analytics.ts
import { apiClient } from "./client";

export interface DashboardKpi {
  total_scans: number;
  last_30d_scans: number;
  unique_cards: number;
  active_members: number;
}

export interface TimelinePoint {
  day: string;
  count: number;
}

export interface RegionStats {
  region: string;
  count: number;
}

export interface DeviceStats {
  device_type: string;
  count: number;
}

export interface TopCard {
  card_id: string;
  last_name: string;
  first_name: string;
  membership_no: string;
  scans: number;
}

export interface DashboardData {
  kpi: DashboardKpi;
  by_day: TimelinePoint[];
  top_regions: RegionStats[];
  top_devices: DeviceStats[];
  top_cards: TopCard[];
}

export interface CardAnalytics {
  card_id: string;
  total_scans: number;
  last_scan: string | null;
  by_day: TimelinePoint[];
  by_region: RegionStats[];
  by_device: DeviceStats[];
}

export interface AnalyticsParams {
  from?: string;
  to?: string;
}

export interface TopActiveUser {
  card_id: string;
  last_name: string;
  first_name: string;
  middle_name?: string | null;
  category_name?: string | null;
  membership_no: string;
  scans: number;
  top_region?: string | null;
  top_device?: string | null;
}

export interface TopActiveUsersResponse {
  items: TopActiveUser[];
  total: number;
  page: number;
  page_size: number;
}

export interface TopActiveParams {
  from?: string;
  to?: string;
  page?: number;
  page_size?: number;
}

export const analyticsApi = {
  dashboard: (params?: AnalyticsParams) =>
    apiClient.get<DashboardData>("/analytics/dashboard", { params }),
  card: (id: string, params?: AnalyticsParams) =>
    apiClient.get<CardAnalytics>(`/analytics/cards/${id}`, { params }),
  topActive: (params?: TopActiveParams) =>
    apiClient.get<TopActiveUsersResponse>("/analytics/top-active", { params }),
  report: (params?: { from?: string; to?: string }) =>
    apiClient.get<Blob>("/analytics/report.xlsx", { params, responseType: "blob" }),
  cardReport: (id: string, params?: { from?: string; to?: string }) =>
    apiClient.get<Blob>(`/analytics/cards/${id}/report.xlsx`, {
      params,
      responseType: "blob",
    }),
};
